File: backend/app/schemas/detection.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import Decision, SensitivityLevel


class SecurityAnalysis(BaseModel):
    similarity_score: float
    facts_matched: int
    facts_total: int
    factual_overlap_score: float
    sensitivity: Optional[SensitivityLevel] = None
    risk_score: int
    decision: Decision
    matched_source: Optional[str] = None
    lineage_tag: Optional[str] = None
    matched_facts: list[str] = Field(default_factory=list)


class DetectRequest(BaseModel):
    generated_response: str = Field(min_length=1)
    document_ids: list[int] = Field(default_factory=list)


class DetectResponse(BaseModel):
    request_id: str
    security_analysis: SecurityAnalysis


class DetectionResultRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    request_id: str
    similarity_score: float
    factual_overlap_score: float
    facts_matched: int
    facts_total: int
    sensitivity_score: int
    risk_score: int
    decision: Decision
    matched_document_id: Optional[int] = None
    matched_facts: list[str]
    created_at: datetime

File: frontend/src/types/detection.ts
export type Decision = 'ALLOW' | 'WARN' | 'BLOCK';

export type SensitivityLevel = 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';

export interface SecurityAnalysis {
  similarity_score: number;
  facts_matched: number;
  facts_total: number;
  factual_overlap_score: number;
  sensitivity?: SensitivityLevel | null;
  risk_score: number;
  risk_level: 'LOW' | 'MEDIUM' | 'HIGH';
  decision: Decision;
  matched_source?: string | null;
  lineage_tag?: string | null;
  matched_facts: string[];
}

export interface GenerateRequest {
  prompt: string;
  scenario?: string;
}

export interface GenerateResponse {
  request_id: string;
  prompt: string;
  generated_response: string;
  security_analysis: SecurityAnalysis;
}

export interface DetectRequest {
  generated_response: string;
  document_ids?: number[];
}

export interface DetectResponse {
  request_id: string;
  security_analysis: SecurityAnalysis;
}
